### PR TITLE
Fix glTF buffer size check

### PR DESCRIPTION
**Pull request by A-Machynak**
A crash (possibly) occurs while importing a glTF file pointing to a buffer with a smaller size than expected (`byteLength` in gltf file is greater than the real size).

This can crash the application and, since the file gets reimported, on the subsequent startups too.
The only way to fix this at the moment is deleting the problematic files from the project folder.

To reproduce the issue:

1. Create new project
2. Import `Gem.gltf` with `buffer.bin` (from `Gem.zip`) into the project

[Gem.zip](https://github.com/godotengine/godot/files/14316860/Gem.zip)

Seems like a simple oversight(?)

### CODE REVIEW

**Comment by lyuma** [comment]
Good find. I think this change is good, this change alone does not seem sufficient to solve the crash for me if I edit buffer byteLength to 844.

I think we *also* need to do the same checks in `GLTFDocument::_parse_buffer_views`.

Also, we should fail if byteLength <=0 . [The gltf spec explicitly forbids byteLength == 0](https://github.com/KhronosGroup/glTF/blob/main/specification/2.0/schema/bufferView.schema.json#L19..L23)

And finally, we should double check the length checks in accessors while we're at it.

**Comment by A-Machynak** [comment]
Sounds good, I'll take a proper look at that when I've got time. I was just trying to figure out why my app is crashing, backtracked to this check and realised that I'm importing some corrupted files. Didn't have time to look at other stuff, so I just pushed a quick fix; though it definitely didn't fix everything.

**Comment by fire** [comment]
@lyuma is this complete?

**Comment by A-Machynak** [comment]
Sorry, quite busy atm with stuff.
Added some bufferview checks mentioned by @lyuma. But the more I look at it, the more I feel like there is a good amount of checks missing and loading some corrupted file is like playing a russian roulette.
I'll be able to get back to it in a few months and see what I can do. Right now, these few simple checks will have to do.

File: modules/gltf/gltf_document.cpp [hunk 1]
```diff
@@ -866,16 +867,31 @@ Error GLTFDocument::_parse_buffer_views(Ref<GLTFState> p_state) {
...
 
 		if (d.has("byteOffset")) {
 			buffer_view->byte_offset = d["byteOffset"];
+			ERR_FAIL_COND_V_MSG(buffer_view->byte_offset < 0, ERR_PARSE_ERROR, "glTF: Bufferview byte_offset < 0 (" + itos(buffer_view->byte_offset) + ")");
 		}
 
 		if (d.has("byteStride")) {
 			buffer_view->byte_stride = d["byteStride"];
+
+			// Magic numbers from gltf bufferview schema
```
Comment (fire) [inline modules/gltf/gltf_document.cpp:887]:
```suggestion
			// Some magic numbers from the gltf bufferview schema.
```

File: modules/gltf/gltf_document.cpp [hunk 2]
```diff
@@ -866,16 +867,31 @@ Error GLTFDocument::_parse_buffer_views(Ref<GLTFState> p_state) {
...
+			ERR_FAIL_COND_V_MSG((buffer_view->byte_stride % 4 != 0) || (buffer_view->byte_stride < 4) || (buffer_view->byte_stride > 252),
+					ERR_PARSE_ERROR, "glTF: Incorrect byte_stride (" + itos(buffer_view->byte_stride) + ")");
 		}
+		// bufferview can't point to something outside the actual data
+		ERR_FAIL_COND_V_MSG((buffer_view->byte_length + buffer_view->byte_offset) > buffersData[buffer_view->buffer].size(),
+				ERR_PARSE_ERROR, "glTF: length+offset > size (" + itos(buffersData[buffer_view->buffer].size()) + ")");
+		// TODO: check the combination of byte_stride, offset and byte_length
```
Comment (fire) [inline modules/gltf/gltf_document.cpp:894]:
```suggestion
		// TODO: Check the combination of byte_stride, offset and byte_length for bounds.
```

**Review by fire — COMMENTED** [review]
I modified some of the comments to match sentence case.

**Comment by lyuma** [comment]
Just checking in a couple months later to see if you have had luck locating any other missing bounds checks. I will make sure we will get this fix in for 4.3 either way, but I always appreciate any help / another pair of eyes

**Comment by A-Machynak** [comment]
Seems fine now. Probably(/hopefully) used an older version while testing or something. I'll find out once I start importing some weird stuff and get annoyed enough to find a fix again.

File: modules/gltf/gltf_document.cpp [hunk 3]
```diff
@@ -857,6 +857,7 @@ Error GLTFDocument::_parse_buffer_views(Ref<GLTFState> p_state) {
 	if (!p_state->json.has("bufferViews")) {
 		return OK;
 	}
+	const Vector<Vector<uint8_t>> &buffersData = p_state->buffers;
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp:860]: Godot style is to use `snake_case` for variable names.

```suggestion
	const Vector<Vector<uint8_t>> &buffers_data = p_state->buffers;
```

File: modules/gltf/gltf_document.cpp [hunk 4]
```diff
@@ -866,16 +867,31 @@ Error GLTFDocument::_parse_buffer_views(Ref<GLTFState> p_state) {
 
 		ERR_FAIL_COND_V(!d.has("buffer"), ERR_PARSE_ERROR);
 		buffer_view->buffer = d["buffer"];
+
+		// Check if the buffer actually exists.
+		// ! This assumes that we've parsed the buffers (p_state->buffers) before!
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp:872]:
```suggestion
		// This assumes that we've parsed the buffers (p_state->buffers) before!
```

File: modules/gltf/gltf_document.cpp [hunk 5]
```diff
@@ -866,16 +867,31 @@ Error GLTFDocument::_parse_buffer_views(Ref<GLTFState> p_state) {
...
+		ERR_FAIL_COND_V_MSG((buffersData.size() < buffer_view->buffer), ERR_PARSE_ERROR,
+				"glTF: Missing buffer at index referenced by bufferview (" + itos(buffer_view->buffer) + ")");
+
 		ERR_FAIL_COND_V(!d.has("byteLength"), ERR_PARSE_ERROR);
 		buffer_view->byte_length = d["byteLength"];
 
 		if (d.has("byteOffset")) {
 			buffer_view->byte_offset = d["byteOffset"];
+			ERR_FAIL_COND_V_MSG(buffer_view->byte_offset < 0, ERR_PARSE_ERROR, "glTF: Bufferview byte_offset < 0 (" + itos(buffer_view->byte_offset) + ")");
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp:881]:
```suggestion
			ERR_FAIL_COND_V_MSG(buffer_view->byte_offset < 0, ERR_PARSE_ERROR, "glTF: Buffer view byte_offset is not allowed to be negative (was " + itos(buffer_view->byte_offset) + ")");
```

File: modules/gltf/gltf_document.cpp [hunk 6]
```diff
@@ -866,16 +867,31 @@ Error GLTFDocument::_parse_buffer_views(Ref<GLTFState> p_state) {
...
 		}
 
 		if (d.has("byteStride")) {
 			buffer_view->byte_stride = d["byteStride"];
+
+			// Some magic numbers from the gltf bufferview schema.
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp:887]:
```suggestion
			// The glTF bufferView schema specifies byteStride, if present, must be a multiple of 4, between 4 and 252.
			// https://github.com/KhronosGroup/glTF/blob/main/specification/2.0/schema/bufferView.schema.json
```

**Pull request closed without merging**